[1.10] Added EnchantmentLevelSetEvent

rockhymas opened the pull request:
Allows enchantment level to be modified before enchantments are chosen
in the enchanting table.

I've got the test mod passing, but I'd like to figure out how to take the forge build I made and test it out with the mod I've built. Figured I would work on that while waiting for this PR to get reviewed. Any pointers on how to do that are welcome, as this is my first foray into forge development.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3373) <br/>All committers have signed the CLA.

howtonotwin commented on patches/minecraft/net/minecraft/inventory/ContainerEnchantment.java.patch:
```diff
@@ -75,3 +75,13 @@
+                         }
++                        net.minecraftforge.event.enchanting.EnchantmentLevelSetEvent e = new net.minecraftforge.event.enchanting.EnchantmentLevelSetEvent(i1, (int)power, itemstack, field_75167_g[i1]);
++                        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e);
++                        field_75167_g[i1] = e.getLevel();
```
This should probably go into `ForgeEventFactory`.

rockhymas replied:
Will do.

howtonotwin commented on src/main/java/net/minecraftforge/event/enchanting/EnchantmentLevelSetEvent.java:
```diff
@@ -0,0 +1,97 @@
+public class EnchantmentLevelSetEvent extends Event
+{
+	private int slot;
+	private int power;
```
Indentation needs to be fixed, and since most of these fields are immutable they should be marked `final`.

rockhymas replied:
yep

liach commented on src/main/java/net/minecraftforge/event/enchanting/EnchantmentLevelSetEvent.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public class EnchantmentLevelSetEvent extends Event
+{
+	private int slot;
```
Slot number means nothing without a container.

rockhymas replied:
These are not container slots, they are enchantment slots. Maybe this should be renamed? It looks like EnchantmentHelper uses the term `enchantNum`, I can do that.

liach commented on src/main/java/net/minecraftforge/event/enchanting/EnchantmentLevelSetEvent.java:
```diff
@@ -0,0 +1,97 @@
+	private int slot;
+	private int power;
+    private ItemStack itemStack;
+    private int level;
```
Add a final field `originalLevel` for listeners to access when `level` is already changed by other listeners.

rockhymas replied:
Good idea.

rockhymas commented:
@howtonotwin updated with your recommendations.

liach reviewed (changes requested):
And providing a world and a position also make the event way more useful.

rockhymas commented:
Updated

liach commented on src/main/java/net/minecraftforge/event/enchanting/EnchantmentLevelSetEvent.java:
```diff
@@ -0,0 +1,135 @@
+     * 
+     * @return the level of the enchantment for this enchantNum (0-30)
+     */
+    public int getLOriginalevel()
```
`getLOriginalevel`

rockhymas replied:
gah! fixed.

rockhymas committed:
Added EnchantmentLevelSetEvent

Allows enchantment level to be modified before enchantments are chosen
in the enchanting table.

Draco18s commented on src/main/java/net/minecraftforge/event/enchanting/EnchantmentLevelSetEvent.java:
```diff
@@ -0,0 +1,135 @@
+{
+    private final World world;
+    private final BlockPos pos;
+    private final int enchantNum;
```
I would change this to "slotNum" as "enchantNum" makes me think it's the ID of the enchantment being applied (but its not).

rockhymas replied:
I was using the term slot before, but it was being confused as having something to do with inventory slots. I agree though, that enchantNum still isn't perfect. What about enchantRow? or enchantLine?

Draco18s replied:
That's a fair point.  "enchantRow" makes sense.

rockhymas committed:
Cleaned up comments and cast from code review

rockhymas committed:
Rename enchantNum to enchantRow

Conflicts:
	src/main/java/net/minecraftforge/event/ForgeEventFactory.java

mezz commented:
Can you explain how you will use this event in your mod?

Also please update this PR to target 1.11. New features are no longer being merged into 1.10.

rockhymas commented:
There is already a 1.11 version of this PR at https://github.com/MinecraftForge/MinecraftForge/issues/3433. I will explain how I plan to use the event over there.

mezz commented:
Thanks. Closing this one since it has been updated to 1.11 there.
